### DISCUSSION TITLE
printf

### DISCUSSION THREAD

**Discussion by oisact** [egzumer/uv-k5-firmware-custom#236] [Ideas]
It appears printf is taking up around 1.5k of the firmware. That is around 2.5% of the firmware is the printf implementation. Looking at the usage of printf this seems very unnecessary.  Its primary usages are in spectrum.c and dtmf.c, and the usages I see can easily be replaced with simpler itoa and strcat in most cases. The dtmf usages are extremely trivial as they are just shorthand for concatenating a string, char and string. It will also perform faster without the printf logic being executed over and over rendering some of those screens.

Some usages, like `sprintf(String, "%s", registerSpecs[idx].name);` don't even make sense as that is just a strcpy.  Or `sprintf(String, "%u", GetRegMenuValue(idx));`  which is just itoa.

**Comment by oisact** [comment]
The arm compiler doesn't have an itoa implementation, so here's one. This can be simplified further if it is known that numbers will never be negative. I'm not familiar enough with the various uses to know that for sure.

```
int itoa(int value, char *sp, int radix)
{
    char tmp[16];// be careful with the length of the buffer
    char *tp = tmp;
    int i;
    unsigned v;

    int sign = (radix == 10 && value < 0);    
    if (sign)
        v = -value;
    else
        v = (unsigned)value;

    while (v || tp == tmp)
    {
        i = v % radix;
        v /= radix;
        if (i < 10)
          *tp++ = i+'0';
        else
          *tp++ = i + 'a' - 10;
    }

    int len = tp - tmp;

    if (sign) 
    {
        *sp++ = '-';
        len++;
    }

    while (tp > tmp)
        *sp++ = *--tp;

    return len;
}
```

Then a convenience function to concatenate a number onto a string:
```
void itoa_cat(int d, char *s, int radix) {
	while (*s != 0) s++;
	itoa(d, s, radix);
}
```

Finally another convenience function to concatenate a character onto a string:
```
void strcatc(char *s, char c) {
	while (*s != 0) s++;
	*s++ = c;
	*s = 0;
}
```

Then we can replace something like this:
```
sprintf(String, "%s%c%s", gEeprom.ANI_DTMF_ID, gEeprom.DTMF_SEPARATE_CODE, gEeprom.KILL_CODE);
```

With:
```
strcpy(String, gEeprom.ANI_DTMF_ID);
strcatc(String, gEeprom.DTMF_SEPARATE_CODE);
strcat(String, gEeprom.KILL_CODE);
```

And replace something like this:
```
sprintf(String, "%d/%d P:%d T:%d", settings.dbMin, settings.dbMax,
          Rssi2DBm(peak.rssi), Rssi2DBm(settings.rssiTriggerLevel));
```

With:
```
itoa(settings.dbMin, String, 10);
strcatc(String, '/');
itoa_cat(settings.dbMax, String, 10);
strcat(String, " P:");
itoa_cat(Rssi2DBm(peak.rssi), String, 10);
strcat(String, " T:");
itoa_cat(Rssi2DBm(settings.rssiTriggerLevel), String, 10);
```

**Comment by egzumer** [comment]
You must have not looked into the code. `sprintf` is used extensively everywhere in everything UI related (94 calls currently). It is huge simplifier in many ways, no way I'm getting rid of it.

**Comment by oisact** [comment]
` It is huge simplifier in many ways, no way I'm getting rid of it.`

Extra baggage in the binaries as a convenience for coding is your prerogative. Personally I work under the mantra that I write code once, and it executes billions of times.

**Reply by egzumer** [reply]
It's not like the radio will suddenly become a speed demon. There is nothing critical going on in the radio. So yeah, I'm lazy bastard and ease of use is more important for me. Maybe I'll have a second look at that, so thank you for pointing that out, but I'm not jumping in changing it right now.

**Reply by oisact** [reply]
My main point was freeing up space in the binary for other more important things. As soon as I saw a printf implementation was included I tried a build with a do-nothing stub, and that freed up 1.5k.  Performance-wise it's inconsequential because all string lengths are tiny. Just pointing out that it is a potential way to include at least another 1k worth of actual functionality in there, and I'm sure you have plenty of ideas for other things you want to include down the road.

**Reply by egzumer** [reply]
Well, we'll see how much longer I will work on this project, it is taxing, right now I'm not onto anything that will critically need more space.
